LIP-7: a composite beacon report receiver

TheDZhon opened the pull request:
Propose a composite beacon report receiver to allow multiple callbacks
invocation per single lido oracle beacon report run.

TheDZhon committed:
feat: add a new LIP spec

Propose a composite beacon report receiver to allow multiple callbacks
invocation per single lido oracle beacon report run.

folkyatina commented on LIPS/lip-composite_beacon_report_receiver.md:
```diff
@@ -0,0 +1,158 @@
+
+## Abstract
+
+The `OrderedCallbacksArray` contract implements the aforementioned `IOrderedCallbacksArray` interface adding the access modifiers to allow storage changing calls (add/insert/remove) only originated by the `Voting` contract (having `require(msg.sender == Voting.address)` ). Another one contract (`CompositePostRebaseBeaconReceiver`) inherits `OrderedCallbacksArray` and implemens `IBeaconReportRecei ...
```
s/Another one contract/Another contract

folkyatina replied:
implemen**t**s

folkyatina replied:
The last sentence can look better being split into smaller, simpler ones.

TheDZhon replied:
Yeah, thx! Fixed.

folkyatina commented on LIPS/lip-composite_beacon_report_receiver.md:
```diff
@@ -0,0 +1,158 @@
+
+## Motivation
+
+Currently, the `LidoOracle` [contract](https://docs.lido.fi/contracts/lido-oracle) provides only one slot for a quorum report intercepting callback (see [LidoOracle.setBeaconReportReceiver](https://docs.lido.fi/contracts/lido-oracle#setbeaconreportreceiver)). We plan to occupy this slot soon with a newly developed contract implementing the [LIP-6 coverage application mechanism](https://research.l ...
```
require **to** append ?

TheDZhon replied:
paraphrased it

arwer13 committed:
Update lip-composite_beacon_report_receiver.md

arwer13 commented on LIPS/lip-composite_beacon_report_receiver.md:
````diff
@@ -0,0 +1,158 @@
+See: `callbacksLength`.
+
+```solidity
+function callbacks(uint256 _atIndex) external view returns (address);
````
Imho, `getCallback` or `callbackAt` would be a clearer name

TheDZhon replied:
Yeah, you are completely right.
But this one is autogenerated getter for the public `callbacks` storage var.
So I just decided to make it consistent from this point of view.

arwer13 commented on LIPS/lip-composite_beacon_report_receiver.md:
```diff
@@ -0,0 +1,158 @@
+Adds the provided `_callback` element at the end of the callbacks array, after its current last element.
+* Reverts if `_callback` address is zero.
+* Reverts if `msg.sender` is not equal to the stored `voting` address.
+* Reverts if the new length of the callbacks array becomes greater than the `MAX_CALLBACKS_CNT` contract-wide constant.
```
Maybe make `uint256 public constant MAX_CALLBACKS_COUNT = 16;` an immutable? To be able to parametrize it on contract creation and to reduce storage usage a bit.

https://github.com/lidofinance/lido-dao/blob/e935ce24cf373f79bdf9d366c714a9cb7ca7e08b/contracts/0.8.9/OrderedCallbacksArray.sol#L17

TheDZhon replied:
Hmm, as I know, both `immutable` and `constant` are occupy zero storage slot space.

I have found also this answer https://ethereum.stackexchange.com/a/113885 which highlights the substitutions difference.

Nevertheless, we set it to `16` just as a sanity check.

arwer13 replied:
Indeed!

arwer13 commented on LIPS/lip-composite_beacon_report_receiver.md:
```diff
@@ -0,0 +1,158 @@
+#### Access control (permissions model)
+There are two permissioned addresses introduced explicitly: `Voting` and `LidoOracle`. Only `Voting` can add/insert/remove callbacks into the `CompositePostReabaseBeaconReceiver` to make it feasible only by the Lido DAO direct vote-backed will. Only `LidoOracle` is allowed to call `processLidoOracleReport` to prevent propagating various fake oracle reports by the added callbacks or execution in a ...
+
+#### Misbehaving callback
```
I propose to note here that if a callback throws an exception all the subsequent callbacks get skipped.

TheDZhon replied:
Good point, I'll add it.

folkyatina commented on LIPS/lip-composite_beacon_report_receiver.md:
````diff
@@ -0,0 +1,158 @@
+```solidity
+function addCallback(address _callback) external override onlyVoting;
+```
+Adds the provided `_callback` element at the end of the callbacks array, after its current last element.
````
Probably, the last part "after its current last element" is excessive.

TheDZhon replied:
Yeah, you are right :+1:

folkyatina reviewed (commented):
Added some remarks on grammar and style, that could be safely ignored in some parts. 

Beside of that, everything is clear. Maybe some comments on chosen data structure can be useful, but on the other hand it seems obvious and can be omitted safely.

TheDZhon committed:
fix: assign lip-7

TheDZhon committed:
fix: post-review fixes by @arwer13 and @folkyatina

Paraphrase some statements.
Explicitly declare that even one callback exception reverts the whole bunch.

TheDZhon committed:
fix: typo
